content/blog/performant-elm-keyed/index.md

```diff
@@ -0,0 +1,155 @@
+nodes have to be compared.
+
+## A working example
+
+Let's unleash our inner Doubting Thomas and test out these assumptions. In
+most UNIX systems there is a dictionary of words in
+`/usr/share/dict/words`, so I've copied more or less 3000 of those and
+wrote a <a href="https://ellie-app.com/6RtqCZC7Bzva1" target="_blank">small
+Ellie app</a> to display them. It looks something like this:
+
+![Ellie example](./ellie-example.png)
+
+To simulate a slighly more complex HTML structure, I've broken each word
+into characters and repeated them twenty times. Also, if you click the
+dropdown at the top of the page, it will toggle between ascending and
+descending sorting.
+
+If we record a performance readout of this example, this is what we'll see:
+
+![Performance before](./before.png)
+
+Let's break down the graph:
+
+- From the moment we click the button, we spend around 1000ms in scripting
+- Then follows 250ms of style recalculation
+- Then follows 2000ms of layout rendering
+- Which is followed by 500ms of layer tree updating
+
+What I would like to focus on is the **right hand side** of the scripting
+portion: you can see that there are a lot of small function calls happening
+at the end. If you zoom in, you will see that the root function
+`_VirtualDom_applyPatches` has instantiated many thousands of times
+`_VirtualDom_applyPatchesHelp`.
+
+## Introducing Html.Keyed
+
+Now let's replace the current rendering logic with `Html.Keyed`. We'll
+replace this line of code:
+
+```elm
+Html.ul [ Attr.class "entries" ]
+    (List.map viewEntry sortedEntries)
+```
+
+with this line of code:
+
+```elm
+Html.Keyed.ul [ Attr.class "entries" ]
+    (List.map viewKeyedEntry sortedEntries)
+```
+
+and implement this new view helper:
+
+```elm
+viewKeyedEntry : Entry -> ( String, Html Msg )
+viewKeyedEntry entry =
+    ( String.fromInt entry.id, viewEntry entry )
+```
+
+You can play around with the <a href="https://ellie-app.com/6RtDQcTkV6ba1" target="_blank">new Ellie app</a>. Now let's make another performance measurement!
+
+![Performance after](./after.png)
+
+As you can see, the general shape of the performance graph hasn't changed.
+The browser is still spending most of the time recalculating styles,
+updating the layout and the layer tree. Unfortunately `Html.Keyed` is not
+black magic and cannot make your application faster if the bottleneck is
+somewhere else.
+
+But if we zoom in the scripting section we'll see an interesting change.
+First of all, it's clear that our browser has been less busy: the usage
+pattern is way less fragmented. And if we take a look at the call stack we
+can see that now `_VirtualDom_applyPatches` calls
+`_VirtualDom_applyPatchesHelp` only once!
+
+## One More Thing
+
+So far we have seen the benefits of `Html.Keyed` as far as it concerns
+performance. But another great use case of this library is **UI consistency**.
+Imagine we wanted to add a little checkbox near each word. Now if I checked
```

Comment:
I find this example a little contrived because the data isn't represented in the model

Reply:
like, ui consistency exists when we're rendering data that's persisted in the model

Reply:
what's an example of an interactive element that doesn't exist in the model (besides our hacky input fixes from elm 18)?